zebra-state/src/service/non_finalized_state.rs

```diff
@@ -294,13 +302,98 @@ impl NonFinalizedState {
+            self.invalidated_blocks.shift_remove_index(0);
+        }
 
         self.update_metrics_for_chains();
         self.update_metrics_bars();
     }
 
+    /// Reconsiders a previously invalidated block and its descendants into the non-finalized state
+    /// based on a block_hash. Reconsidered blocks are inserted into the previous chain and re-inserted
+    /// into the chain_set.
+    pub fn reconsider_block(&mut self, block_hash: block::Hash) -> Result<(), ReconsiderError> {
+        // Get the invalidated blocks that were invalidated by the given block_hash
+        let height = self
+            .invalidated_blocks
+            .iter()
+            .find_map(|(height, blocks)| {
+                if blocks.first()?.hash == block_hash {
+                    Some(height)
+                } else {
+                    None
+                }
+            })
+            .ok_or(ReconsiderError::MissingInvalidatedBlock(block_hash))?;
+
+        let mut invalidated_blocks = self
+            .invalidated_blocks
+            .clone()
+            .shift_remove(height)
+            .ok_or(ReconsiderError::MissingInvalidatedBlock(block_hash))?;
+        let mut_blocks = Arc::make_mut(&mut invalidated_blocks);
+
+        // Find and fork the parent chain of the invalidated_root. Update the parent chain
+        // with the invalidated_descendants
+        let invalidated_root = mut_blocks
+            .first()
+            .ok_or(ReconsiderError::InvalidatedBlocksEmpty)?;
+
+        let root_parent_hash = invalidated_root.block.header.previous_block_hash;
+        let parent_chain = self
+            .parent_chain(root_parent_hash)
+            .map_err(|_| ReconsiderError::ParentChainNotFound(block_hash));
+
+        let modified_chain = match parent_chain {
+            // 1. If a parent chain exist use the parent chain
+            Ok(parent_chain) => {
+                let mut chain = Arc::unwrap_or_clone(parent_chain);
+
+                for block in Arc::unwrap_or_clone(invalidated_blocks) {
+                    chain = chain.push(block)?;
+                }
+
+                Arc::new(chain)
+            }
+            // 2. If a parent chain does not exist create a new chain from the non finalized state tip
+            Err(_) => {
+                let mut chain = Chain::new(
+                    &self.network,
+                    invalidated_root.height.previous().map_err(|_| {
+                        ReconsiderError::InvalidHeight(Height(invalidated_root.height.0 - 1))
+                    })?,
+                    Default::default(),
+                    Default::default(),
+                    Default::default(),
+                    Default::default(),
+                    ValueBalance::zero(),
+                );
+
+                for block in Arc::unwrap_or_clone(invalidated_blocks) {
+                    chain = chain.push(block)?;
+                }
+
+                Arc::new(chain)
+            }
+        };
```

Comment:
`parent_chain()` will already return a new fork if the invalidated block's parent is in the non-finalized state. This method shouldn't add a new chain to the non-finalized state if it's root is below the finalized tip.

This method could accept a handle to the db as a parameter and create a new chain if the invalidated block's parent is the finalized tip block, or it could refuse to reconsider blocks unless their parent is in the non-finalized state.

The suggestion below does the latter:

```suggestion
        let parent_chain = self
            .parent_chain(root_parent_hash)
            .map_err(|_| ReconsiderError::ParentChainNotFound(block_hash))?;

        let mut modified_chain = Arc::unwrap_or_clone(parent_chain);

        for block in Arc::unwrap_or_clone(invalidated_blocks) {
            modified_chain = modified_chain.push(block)?;
        }
```